src/pycapsule.rs

```diff
@@ -0,0 +1,296 @@
+///
+///
+/// # Example
+/// ```
+///  use std::ffi::CString;
+///  use pyo3::{prelude::*, pycapsule::PyCapsule};
+///
+///  #[repr(C)]
+///  struct Foo {
+///     pub val: u32,
+///  }
+///
+///  let r = Python::with_gil(|py| -> PyResult<()> {
+///     let foo = Foo { val: 123 };
+///     let name = CString::new("builtins.capsule").unwrap();
+///
+///     let capsule = PyCapsule::new(py, foo, name.as_ref())?;
+///
+///     let module = PyModule::import(py, "builtins")?;
+///     module.add("capsule", capsule)?;
+///
+///     let cap: &Foo = unsafe { PyCapsule::import(py, name.as_ref(), false)? };
+///     assert_eq!(cap.val, 123);
+///     Ok(())
+///  });
+///  assert!(r.is_ok());
+/// ```
+#[repr(transparent)]
+pub struct PyCapsule(PyAny);
+
+pyobject_native_type_core!(PyCapsule, ffi::PyCapsule_Type, #checkfunction=ffi::PyCapsule_CheckExact);
+
+impl PyCapsule {
+    /// Constructs a new capsule of whose contents are `T` associated with `name`.
+    pub fn new<'py, T: 'static>(py: Python<'py>, value: T, name: &CStr) -> PyResult<&'py Self> {
+        Self::new_with_destructor(py, value, name, std::mem::drop)
+    }
+
+    /// Constructs a new capsule of whose contents are `T` associated with `name`
+    /// Provide a destructor for when `PyCapsule` is destroyed it will be passed the capsule.
+    pub fn new_with_destructor<'py, T: 'static, F: FnOnce(T)>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+        destructor: F,
+    ) -> PyResult<&'py Self> {
+        let val = Box::new(CapsuleContents { value, destructor });
+
+        let cap_ptr = unsafe {
+            ffi::PyCapsule_New(
+                Box::into_raw(val) as *mut c_void,
+                name.as_ptr(),
+                Some(capsule_destructor::<T, F>),
+            )
+        };
+        if cap_ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(unsafe { py.from_owned_ptr::<PyCapsule>(cap_ptr) })
+        }
+    }
+
+    /// Import an existing capsule.
+    ///
+    /// The `name` should match the path to module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the
+    /// capsule. `no_block` indicates to use
+    /// [PyImport_ImportModuleNoBlock()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModuleNoBlock)
+    /// or [PyImport_ImportModule()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModule)
+    /// when accessing the capsule.
+    ///
+    /// # Safety
+    ///
+    /// It must be known that this capsule's pointer is to an item of type `T`.
+    pub unsafe fn import<'py, T>(py: Python<'py>, name: &CStr, no_block: bool) -> PyResult<&'py T> {
+        let ptr = ffi::PyCapsule_Import(name.as_ptr(), no_block as c_int);
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(&*(ptr as *const T))
+        }
+    }
+
+    /// Set a context pointer in the capsule to `T`
```

Comment:
```suggestion
    /// Sets the context pointer in the capsule to `T`.
```

This Box is leaked; there is no place where the context is destructed. I guess this is unavoidable, but it must be at least be made explicit in the docstring.

Also, either needs a `T: Send` or `get_context` must make sure not to be called from a different thread (but that would limit its usability severely).

Also needs a little explanation of what the context is, and that it's different from `value`.

Reply:
Took a swing at it, hard to find any information on exactly what context is used for itself, so I at least explains what I think it is used for. :sweat_smile: 

Maybe we can alter `capsule_destructor` to pass the whole of `PyCapsule`  (un borrowed) to the destructor, then add `take_value` / `take_context`  methods which would, in my mind at least, allow the user supplied `destructor` to be `Fn(PyCapsule)` and thus full ownership to cleanup anything?

Reply:
Ah, good point. The nice thing about the `FnOnce(T)` design we've currently got is that the user doesn't have to do any `unsafe` type casting in their destructor implementation.

> full ownership to cleanup anything

We can't actually have unborred `PyCapsule` as we don't know its true memory layout (that's private to CPython).

We could potentially have `FnOnce(T, *mut void)` and let the user handle non-null context? That's a bit icky, however I don't see much better. `new_with_context` constructor? Then we'll have combinatorial explosion of constructors :(

Reply:
I see, for now, I've went with your last suggestion (`FnOnce(T, *mut void)`)